{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1c484f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Datasets of this study can be download from  http://spatial.libd.org/spatialLIBD/\n",
    "#https://github.com/QihuangZhang/CeLEry/blob/main/code_paper/1_LIBD/preprocess.py\n",
    "\n",
    "\n",
    "import os,csv,re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import math\n",
    "from skimage import io, color\n",
    "\n",
    "from scipy.sparse import issparse\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import matplotlib.colors as clr\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "#Read original data and save it to h5ad\n",
    "from scanpy import read_10x_h5\n",
    "import CeLEry as cel\n",
    "import json\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b275a9b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### ------------------------------------------------------------------------------------------------------- ###\n",
    "###        Process the genelist\n",
    "### ------------------------------------------------------------------------------------------------------- ###\n",
    "def get_LIBD_top_DEgenes (studyID):\n",
    "    \"\"\"\n",
    "    Preprocess the spatial transcriptomic raw data and obtain the optimal DE genes\n",
    "    Parameters\n",
    "        -----------\n",
    "        studyID : string. the study ID of the LIBD datasets\n",
    "    Returns\n",
    "    -----------\n",
    "        gene_topDE_list: the list of gene set that contains the highest DE genes between layers\n",
    "    \"\"\"\n",
    "    adata = read_10x_h5(\"{studyID}/{studyID}_raw_feature_bc_matrix.h5\".format(studyID = studyID))\n",
    "    spatial = pd.read_csv(\"{studyID}_tissue_positions_list.txt\".format(studyID = studyID),sep=\",\", header = None, na_filter = False, index_col = 0) \n",
    "    adata.obs[\"x1\"] = spatial[1]\n",
    "    adata.obs[\"x2\"] = spatial[2]\n",
    "    adata.obs[\"x3\"] = spatial[3]\n",
    "    # Select captured samples\n",
    "    adata = adata[adata.obs[\"x1\"] == 1]\n",
    "    adata.var_names = [i.upper() for i in list(adata.var_names)]\n",
    "    adata.var[\"genename\"] = adata.var.index.astype(\"str\")\n",
    "    #\n",
    "    del adata.obs[\"x1\"]\n",
    "    #\n",
    "    adata.obs[\"Layer\"] = 0\n",
    "    LayerName =[\"L1\",\"L2\",\"L3\",\"L4\",\"L5\",\"L6\",\"WM\"] #\n",
    "    for i in range(7):\n",
    "        Layer = pd.read_csv(\"{studyID}/{studyID}_{Lname}_barcodes.txt\".format(studyID = studyID, Lname = LayerName[i]), sep=\",\", header = None, na_filter = False, index_col = 0)\n",
    "        adata.obs.loc[Layer.index, \"Layer\"] = int(i+1)\n",
    "        adata.obs.loc[Layer.index, \"Layer_character\"] = LayerName[i]\n",
    "    data = adata[adata.obs[\"Layer\"]!=0]    # Newly added on May 25 #Remove the spots without any layer label\n",
    "    #\n",
    "    #  Preprocessing\n",
    "    adata.var_names_make_unique()\n",
    "    cel.prefilter_genes(adata,min_cells=3) # avoiding all genes are zeros\n",
    "    cel.prefilter_specialgenes(adata)\n",
    "    #Normalize and take log for UMI-------\n",
    "    sc.pp.normalize_per_cell(adata)\n",
    "    sc.pp.log1p(adata)\n",
    "    #\n",
    "    sc.tl.rank_genes_groups(adata, 'Layer_character', method = 'wilcoxon', key_added = \"wilcoxon\")\n",
    "    # sc.pl.rank_genes_groups(adata, n_genes = 200, sharey = False, key=\"wilcoxon\", save = '{studyID}.pdf'.format(studyID = studyID))\n",
    "    gene_topDE_list = []\n",
    "    for layer_i in ['L1', 'L2', 'L3', 'L4', 'L5', 'L6', 'WM']:\n",
    "        gene_rank = sc.get.rank_genes_groups_df (adata,  group = layer_i, key = 'wilcoxon')\n",
    "        top_gene_list = list(gene_rank[\"names\"].iloc[0:200])\n",
    "        gene_topDE_list.append( top_gene_list  )\n",
    "    return gene_topDE_list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cdeebd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist73 = get_LIBD_top_DEgenes (151673)\n",
    "genelist74 = get_LIBD_top_DEgenes (151674)\n",
    "genelist75 = get_LIBD_top_DEgenes (151675)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9afea067",
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist76 = get_LIBD_top_DEgenes (151676)\n",
    "genelist07 = get_LIBD_top_DEgenes (151507)\n",
    "genelist08 = get_LIBD_top_DEgenes (151508)\n",
    "genelist09 = get_LIBD_top_DEgenes (151509)\n",
    "genelist10 = get_LIBD_top_DEgenes (151510)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93b7a703",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the gene list from the pre-screening\n",
    "genelistlist = genelist73 + genelist74 + genelist75 + genelist76 + genelist07 + genelist08 + genelist09 + genelist10\n",
    "genelist = sum(genelistlist, [])  # merge the list of lists\n",
    "genelistuni = list( dict.fromkeys(genelist) )   # remove duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6bde254",
   "metadata": {},
   "outputs": [],
   "source": [
    "def Preprocess_SpTrans (studyID):\n",
    "    \"\"\"\n",
    "    Preprocess the spatial transcriptomic raw data and obtain the optimal DE genes\n",
    "    Parameters:\n",
    "        studyID (string): the study ID of the LIBD datasets\n",
    "    Returns:\n",
    "        gene_topDE_list (list): the list of gene set that contains the highest DE genes between layers\n",
    "    \"\"\"\n",
    "    # Read in the spatial transcriptomic data from a 10x Genomics-formatted HDF5 file\n",
    "    adata = sc.read_10x_h5(f\"{studyID}/{studyID}_raw_feature_bc_matrix.h5\")\n",
    "    \n",
    "    # Read in the spatial coordinates of the tissue samples from a CSV file\n",
    "    spatial = pd.read_csv(f\"{studyID}/{studyID}_tissue_positions_list.txt\", sep=\",\", header=None, na_filter=False, index_col=0) \n",
    "    \n",
    "    # Add the spatial coordinates to the data object\n",
    "    adata.obs[\"x1\"] = spatial[1]\n",
    "    adata.obs[\"x2\"] = spatial[2]\n",
    "    adata.obs[\"x3\"] = spatial[3]\n",
    "    \n",
    "    # Select only the samples that were captured\n",
    "    adata = adata[adata.obs[\"x1\"] == 1]\n",
    "    \n",
    "    # Ensure that all gene names are uppercase\n",
    "    adata.var_names = [i.upper() for i in list(adata.var_names)]\n",
    "    \n",
    "    # Add the gene names as a column in the `var` attribute of the `adata` object\n",
    "    adata.var[\"genename\"] = adata.var.index.astype(\"str\")\n",
    "    \n",
    "    # Remove the \"x1\" column from the `obs` attribute of the `adata` object\n",
    "    del adata.obs[\"x1\"]\n",
    "    \n",
    "    # Add a \"Layer\" column to the `obs` attribute of the `adata` object and initialize it to 0\n",
    "    adata.obs[\"Layer\"] = 0\n",
    "    \n",
    "    # Define a list of layer names\n",
    "    LayerName =[\"L1\", \"L2\", \"L3\", \"L4\", \"L5\", \"L6\", \"WM\"]\n",
    "    \n",
    "    # Loop through each layer and add the corresponding layer number to the `Layer` column of the `adata` object\n",
    "    # Also add the layer name as a separate column for convenience\n",
    "    for i in range(7):\n",
    "        Layer = pd.read_csv(f\"{studyID}/{studyID}_{LayerName[i]}_barcodes.txt\", sep=\",\", header=None, na_filter=False, index_col=0)\n",
    "        adata.obs.loc[Layer.index, \"Layer\"] = int(i+1)\n",
    "        adata.obs.loc[Layer.index, \"Layer_character\"] = LayerName[i]\n",
    "    \n",
    "    # Remove the spots without any layer label\n",
    "    adata = adata[adata.obs[\"Layer\"] != 0]\n",
    "    \n",
    "    # Make the gene names unique\n",
    "    adata.var_names_make_unique()\n",
    "    \n",
    "    # Filter out genes that are expressed in fewer than 3 cells\n",
    "    cel.prefilter_genes(adata, min_cells=3)\n",
    "    \n",
    "    # Filter out special genes (e.g. mitochondrial genes)\n",
    "    cel.prefilter_specialgenes(adata)\n",
    "    \n",
    "    # Normalize the data by cell and take the log of the UMI counts\n",
    "    sc.pp.normalize_per_cell(adata)\n",
    "    sc.pp.log1p(adata)\n",
    "    #\n",
    "    sc.tl.rank_genes_groups(adata, 'Layer_character', method = 'wilcoxon', key_added = \"wilcoxon\")\n",
    "    sc.pl.rank_genes_groups(adata, n_genes = 50, sharey = False, key=\"wilcoxon\", save = '{studyID}.pdf'.format(studyID = studyID))\n",
    "    #\n",
    "    #  Filter the Genes that are selected by SpaGCN\n",
    "    genename = adata.var['genename']\n",
    "    genelistindex = [genename[genename == i].index[0] for i in genelistuni if  len(genename[genename == i])>0]  # only keep the genes that exists in SpT data\n",
    "    # Filter the genelist and output the results\n",
    "    bdata = adata[:,genelistindex]\n",
    "    cdata = sc.AnnData(X = bdata.X.toarray(), obs = bdata.obs, var = bdata.var, uns =bdata.uns, obsm = bdata.obsm)\n",
    "    cdata.write_h5ad(\"data_{studyID}.h5ad\".format(studyID = studyID))\n",
    "    return genelistindex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7983ba04",
   "metadata": {},
   "outputs": [],
   "source": [
    "genelistuni = Preprocess_SpTrans(151673)\n",
    "genelistuni = Preprocess_SpTrans(151674)\n",
    "genelistuni = Preprocess_SpTrans(151675)\n",
    "genelistuni = Preprocess_SpTrans(151676)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8b9f00e",
   "metadata": {},
   "outputs": [],
   "source": [
    "genelistuni = Preprocess_SpTrans(151507)\n",
    "genelistuni = Preprocess_SpTrans(151508)\n",
    "genelistuni = Preprocess_SpTrans(151509)\n",
    "genelistuni = Preprocess_SpTrans(151510)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
